### PR TITLE
[MO args][ONNX FE]fix cutting graph with input, output or both

### PR DESCRIPTION
**Pull request by bszmelcz**
### Details:
 - add function to extract tensor names in order to cut graph
 - [Draft PR](https://github.com/intel-innersource/frameworks.ai.openvino.tests/pull/260) with tests covering different configurations of cutting graph with new FE
### Tickets:
 - 71902

### CODE REVIEW

**Review by tomdol — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: done

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: done

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (bszmelcz) [reply tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py]: done

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (mbencer) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:66]: Can we avoid passing this `node_pre` bool? Maybe enum: `operation_consuming`/`operation_producing` or even string

**Review by tomdol — APPROVED** [review]
Using enum instead of bool is something that I suggested too offline. Overall looks good but we should still wait for a review from @nosovmik

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (nosovmik) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:39]: Is my understanding correct that if we have tensor with name 'myTensor', then `try_get_tensor("myTensor:12345")` will always return this tensor? Is it expected behavior?

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:86]: Getting 2 times 'tensor name' and 3 times 'operation name' just for 'operation:7' looks too much. If number of calls can't be reduced, I'd suggest to add some comment about why so much calls are expected

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:39]: My understanding is that we either expect that there is tensor with name "myTensor:12345" or operation with name "myTensor" and we want its output port 12345. At the moment if we have tensor with name "myTensor" then it will no be returned if we use ```try_get_tensor("myTensor:12345")``` . The function will return tensor only if there is an tensor with name " myTensor:12345" or operation with name "myTensor" and output port 12345. Is it expected behaviour?

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (bszmelcz) [reply tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py]: I was able to reduce the number of function calls

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: added enum

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: I have removed it in the previous commit

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:87]: I checked - 'get_place_by_operation_name' for `operation` is called twice from MO now which looks a bit redundant. Originally test was intended to verify that there is no redundant calls to frontend are happened from MO side.
If it is not possible to avoid redundant calls now, let's set just `assert model_stat.get_place_by_operation_name > 1`

Comment (bszmelcz) [reply tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:87]: There is no way to reduce the number of function calls to 1 in every possible scenario, but I was able to achieve it for the majority of the available combinations of node names

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: done

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py]: done

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (nosovmik) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:26]: I believe this ':' is not needed - it is covered in regexp

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:112]: In case if someone in future makes changes and avoids redundant call - at lest there will be no need to edit this test back to == 1
```suggestion
        assert model_stat.get_place_by_tensor_name >= 1
```

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:125]:
```suggestion
        assert model_stat.get_place_by_tensor_name >= 1
```

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:138]:
```suggestion
        assert model_stat.get_place_by_operation_name >= 1
```

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:151]:
```suggestion
        assert model_stat.get_place_by_operation_name >= 1
```

File: tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py
Comment (nosovmik) [inline tools/mo/unit_tests/mo/moc_frontend/moc_extractor_test_actual.py:201]:
```suggestion
        assert model_stat.get_place_by_operation_name >= 1
```

**Review by nosovmik — APPROVED** [review]
Generally looks good to me, few minor comments

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:26]: It is not needed in general, but I added it to avoid 2 regex calls if there is no ":" in the node name but sure can be removed

**Review by artkuli — APPROVED** [review]

**Review by mbencer — APPROVED** [review]

File: tools/mo/openvino/tools/mo/moc_frontend/pipeline.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/pipeline.py:101]: We have to update the names, because on the second iteration of `decode_name_with_port` called in pipeline.py, after cutting the input, when inputs are not equal, there is no longer node with the name `node_name`. As a basic example in the mnist-8.onnx model we have the node with name `ReLU114`. If we use `mo.py` for this model with the flag `
--input=ReLU114` in the first iteration of calling decode_with_port_name (before cutting the graph) the node with name `ReLU114` is found, but in the second iteration when it becomes the input of the graph its name is changed to `ReLU114_Output_0` (its output tensor), so if we do not update the node name in this line, we will get the error "[ ERROR ]  No node with name ReLU114" which was the source of ticket for this PR

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:28]: sure, agree

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:40]: Indeed, this part makes little sense, I will address that

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (rkazants) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:62]: what if port is `None`? what input port is retrieved then?

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (rkazants) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:54]: why do you try to search tensor with name without port? For example, in case "name:0" you search a tensor with name "name". Is it valid approach?

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (rkazants) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:85]: please remove commented out line

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (rkazants) [inline tools/mo/openvino/tools/mo/moc_frontend/extractor.py:100]: I think there are some duplicating code for pre match and post match cases. We can generalize it

**Comment by rkazants** [comment]
@bszmelcz, please address some comments. Also, I think it makes sense to run DLB for your PR. Do you know how many models of ONNX, PDPD are affected by your PR?

**Comment by bszmelcz** [comment]
> 
In general cutting graphs with the new fe does not work, so this PR fixes cutting graphs in general with the new fe

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:54]: I thought it is expected, if not I can easily remove it

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:100]: ok, I will try to generalize it better

File: tools/mo/openvino/tools/mo/moc_frontend/extractor.py
Comment (bszmelcz) [reply tools/mo/openvino/tools/mo/moc_frontend/extractor.py:62]: if line 62 is being called, then port cannot be None, since to get there there must be regex match with `node_name` which extracts port:name

**Review by rkazants — COMMENTED** [review]
I am fine with these changes. Let me check if PDPD supports cutting feature

**Comment by rkazants** [comment]
@ceciliapeng2011, could you please take a look at this change? The model cutting feature is a bit changed, can it affect PDPD models?

**Comment by ceciliapeng2011** [comment]
@nosovmik Do you have unit tests for PDPD to test graph cutting feature? Is there any dependency to paddle-fe (c++ part) because of this PR?

**Comment by nosovmik** [comment]
> @nosovmik Do you have unit tests for PDPD to test graph cutting feature? Is there any dependency to paddle-fe (c++ part) because of this PR?

Yes, paddle-fe supports graph cutting feature and there are tests for this. Changes here don't affect cutting logic much - there are additional calls of `get_place_by_tensor_name` and `get_place_by_operation_name` which is fine. Paddle FE already supports `get_place_by_tensor_name` properly (returns nullptr if not found) and uses default implementation of `get_place_by_operation_name` which is also correct

**Comment by rkazants** [comment]
> > @nosovmik Do you have unit tests for PDPD to test graph cutting feature? Is there any dependency to paddle-fe (c++ part) because of this PR?
> 
> Yes, paddle-fe supports graph cutting feature and there are tests for this. Changes here don't affect cutting logic much - there are additional calls of `get_place_by_tensor_name` and `get_place_by_operation_name` which is fine. Paddle FE already supports `get_place_by_tensor_name` properly (returns nullptr if not found) and uses default implementation of `get_place_by_operation_name` which is also correct

@nosovmik, @ceciliapeng2011, are there any real models requiring cutting feature? Or just unit tests?

**Comment by mlukasze** [comment]
@rkazants @nosovmik @ceciliapeng2011 
can we merge it if you have no additional change requests?

**Comment by ceciliapeng2011** [comment]
> 

as far as I know, there is no real paddle model requires graph cutting so far.

**Review by ceciliapeng2011 — APPROVED** [review]

**Comment by rkazants** [comment]
> @rkazants @nosovmik @ceciliapeng2011 can we merge it if you have no additional change requests?

Merged!

**Comment by mlukasze** [comment]
> > @rkazants @nosovmik @ceciliapeng2011 can we merge it if you have no additional change requests?
> 
> Merged!

awesome, thanks all!
"see you soon" ;)

**Pull request merged**
